src/Aspire.Hosting/ExecutableResourceBuilderExtensions.cs

```diff
@@ -69,30 +79,64 @@ public static IResourceBuilder<ExecutableResource> AddExecutable(this IDistribut
+        return builder.PublishAsDockerFile(c =>
+        {
+            foreach (var arg in buildArgs ?? [])
+            {
+                c.WithBuildArg(arg.Name, arg.Value);
+            }
+        });
     }
 
-    private static async Task WriteExecutableAsDockerfileResourceAsync(ManifestPublishingContext context, ExecutableResource executable, IEnumerable<DockerBuildArg>? buildArgs = null)
+    /// <summary>
+    /// Adds support for containerizating this <see cref="ExecutableResource"/> during deployment.
+    /// The resulting container image is built, and when the optional <paramref name="configure"/> action is provided,
+    /// it is used to configure the container resource.
+    /// </summary>
+    /// <typeparam name="T">Type of executable resource</typeparam>
+    /// <param name="builder">Resource builder</param>
+    /// <param name="configure">Optional action to configure the container resource</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<T> PublishAsDockerFile<T>(this IResourceBuilder<T> builder, Action<IResourceBuilder<ContainerResource>>? configure)
+        where T : ExecutableResource
     {
-        context.Writer.WriteString("type", "dockerfile.v0");
+        if (!builder.ApplicationBuilder.ExecutionContext.IsPublishMode)
+        {
+            return builder;
+        }
 
-        var appHostRelativePathToDockerfile = Path.Combine(executable.WorkingDirectory, "Dockerfile");
-        var manifestFileRelativePathToDockerfile = context.GetManifestRelativePath(appHostRelativePathToDockerfile);
-        context.Writer.WriteString("path", manifestFileRelativePathToDockerfile);
+        // The implementation here is less than ideal, but we don't have a clean way of building resource types
+        // that change their behavior based on the context. In this case, we want to change the behavior of the
+        // resource from an ExecutableResource to a ContainerResource. We do this by removing the ExecutableResource
+        // from the application model and adding a new ContainerResource in its place in publish mode.
 
-        var manifestFileRelativePathToContextDirectory = context.GetManifestRelativePath(executable.WorkingDirectory);
-        context.Writer.WriteString("context", manifestFileRelativePathToContextDirectory);
+        // There are still dangling references to the original ExecutableResource in the application model, but
+        // in publish mode, it won't be used. This is a limitation of the current design.
+        builder.ApplicationBuilder.Resources.Remove(builder.Resource);
 
-        if (buildArgs is not null)
-        {
-            context.WriteDockerBuildArgs(buildArgs);
-        }
+        var container = new ExecutableContainerResource(builder.Resource);
+        var cb = builder.ApplicationBuilder.AddResource(container);
+        cb.WithImage(builder.Resource.Name);
```

Comment:
For my understanding, why do we need to call `WithImage` when we call `WithDockerfile`? Is that just to name the image that will be produced?

Reply:
Also, doesn't `WithDockerfile` just overwrite what we pass here?

https://github.com/dotnet/aspire/blob/882ae2be3e02edbfc1d222ee97fe2da4eea3fe99/src/Aspire.Hosting/ContainerResourceBuilderExtensions.cs#L366-L371

Reply:
Good catch!

Reply:
Nah you need aan image name.

Reply:
System.InvalidOperationException : The resource 'foo' does not have a container image specified. Use WithImage to specify the container image and tag.

Reply:
ah - it's because of the ORDER of the calls above. If you change it to:

```C#
        return builder.WithAnnotation(annotation, ResourceAnnotationMutationBehavior.Replace)
                      .WithImage(imageName)
                      .WithImageRegistry(registry: null)
                      .WithImageTag("latest");
```

it will work. `WithImageRegistry` is throwing because it can't find the `ContainerImageAnnotation`.

Reply:
I added a comment

Reply:
I think we should fix WithDockerfile.

The current proposed code is confusing because you are going to get a different Image name that what is being specified here - since WithDockerfile is going to overwrite it.

Reply:
not in this PR

Reply:
I still think the code you are adding in this PR is confusing. But I'm not going to block on it.